qiime2/core/type/primitive.py

```diff
@@ -419,12 +419,12 @@ def validate_field(self, name, field):
             raise TypeError("Unsupported type in field: %r"
                             % (field.get_name(),))
 
-    def decode(self, metadata):
+    def decode(self, value):
         # This interface should have already retrieved this object.
-        if not self.is_element(metadata):
+        if not isinstance(value, metadata.MetadataColumn):
```

Comment:
@thermokarst, realized I was overthinking it, this change gets 99% of the intention of the original implementation for basically no additional work.

Normal type-checking will handle the rest.

Reply:
:sake: LGTM!